api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java

```diff
@@ -740,7 +755,292 @@ public void shouldTestPurgeProgramAttributeType() throws Exception {
                 int totalAttributeTypes = pws.getAllProgramAttributeTypes().size();
 		pws.purgeProgramAttributeType(programAttributeType);
 		assertEquals((totalAttributeTypes - 1), pws.getAllProgramAttributeTypes().size());
-}
+	}
+	
+	
+	@Test
+	public void getPrograms_shouldTestGetPrograms() {
+		List<Program> malPrograms = pws.getPrograms("MAL");
+		List<Program> prPrograms = pws.getPrograms("PR");
+		assertEquals(malPrograms.size(), 1);
+		assertEquals(prPrograms.size(), 3);
+	}
+	
+	@Test
+	public void retireProgram_shouldSetRetiredStateToFalseAndSetAReason() {
+		executeDataSet("org/openmrs/api/include/ProgramWorkflowServiceTest-otherProgramWorkflows.xml");
+		Concept concept = Context.getConceptService().getConcept(12);
+		ProgramWorkflow programWorkflow = dao.getProgramWorkflowsByConcept(concept).get(1);
+		ArrayList<ProgramWorkflow> programWorkflows = new ArrayList<>();
+		programWorkflows.add(programWorkflow);
+		Set<ProgramWorkflow> allWorkflows = new HashSet<>(programWorkflows);
+		Program programTest = pws.getAllPrograms().get(1);
+		programTest.setAllWorkflows(allWorkflows);
+		Program program = pws.retireProgram(programTest, "expired");
+		for (ProgramWorkflow workflow : program.getWorkflows()) {
+			assertEquals(workflow.getRetired(), true);
+			for (ProgramWorkflowState state : workflow.getStates()) {
+				assertNotNull(workflow.getStates());
+				assertEquals(state.getRetired(), true);
+				assertEquals(state.getRetired(), true);
+			}
+		}
+		assertEquals(program.getRetireReason(), "expired");
+	}
+	
+	@Test
+	public void unretireProgram_shouldSetRetireFalseForWorkflowsAndWorkflowStates() {
+		Program program = pws.getAllPrograms().get(0);
+		Date lastModifiedDate = program.getDateChanged();
+		Program program1 = pws.unretireProgram(program);
+		assertEquals(program.getRetired(), false);
+		for (ProgramWorkflow workflow : program.getAllWorkflows()) {
+			if (lastModifiedDate != null && lastModifiedDate.equals(workflow.getDateChanged())) {
+				assertEquals(workflow.getRetired(), false);
+				for (ProgramWorkflowState state : workflow.getStates()) {
+					if (lastModifiedDate.equals(state.getDateChanged())) {
+						assertEquals(state.getRetired(), false);
+					}
+				}
+			}
+		}
+	}
+	
+	@Test
+	public void saveUnretireProgram_shouldTestSaveUnretireProgram() {
+		Program program = new Program();
+		Concept concept = Context.getConceptService().getAllConcepts().get(0);
+		User testUser = Context.getUserService().getAllUsers().get(0);
+		program.setConcept(concept);
+		program.setDescription("test");
+		program.setName("programTest");
+		program.setCreator(testUser);
+		pws.getAllPrograms();
+		pws.retireProgram(program, "abc");
+		pws.unretireProgram(program);
+		Program program1 = dao.getProgramByUuid(program.getUuid());
```

Comment:
Please think of a better variable name
"program1" doesn't really signify anything.

Reply:
I changed the name